PhysicsTools/NanoAOD/plugins/LeptonJetVarProducer.cc

```diff
@@ -0,0 +1,211 @@
+};
+
+//
+// constants, enums and typedefs
+//
+
+
+//
+// static data member definitions
+//
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+template <typename T>
+void
+LeptonJetVarProducer<T>::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const
+{
+
+  edm::Handle<edm::View<pat::Jet>> srcJet;
+  iEvent.getByToken(srcJet_, srcJet);
+  edm::Handle<edm::View<T>> srcLep;
+  iEvent.getByToken(srcLep_, srcLep);
+  edm::Handle<std::vector<reco::Vertex>> srcVtx;
+  iEvent.getByToken(srcVtx_, srcVtx);
+
+  unsigned nJet = srcJet->size();
+  unsigned nLep = srcLep->size();
+
+  std::vector<float> ptRatio(nLep,-1);
+  std::vector<float> ptRel(nLep,-1);
+  std::vector<float> jetNDauChargedMVASel(nLep,0);
+  std::vector<reco::CandidatePtr> jetForLepJetVar(nLep,reco::CandidatePtr());
+
+  const auto & pv = (*srcVtx)[0];
+
+  for (uint il = 0; il<nLep; il++){
+    for (uint ij = 0; ij<nJet; ij++){
+      auto lep = srcLep->ptrAt(il);
+      auto jet = srcJet->ptrAt(ij);
+      if(matchByCommonSourceCandidatePtr(*lep,*jet)){
+	  auto res = calculatePtRatioRel(lep,jet,pv);
+	  ptRatio[il] = std::get<0>(res);
+	  ptRel[il] = std::get<1>(res);
+	  jetNDauChargedMVASel[il] = std::get<2>(res);
+	  jetForLepJetVar[il] = jet;
+	  break; // take leading jet with shared source candidates
+	}
+    }
+  }
+
+  std::unique_ptr<edm::ValueMap<float>> ptRatioV(new edm::ValueMap<float>());
+  edm::ValueMap<float>::Filler fillerRatio(*ptRatioV);
+  fillerRatio.insert(srcLep,ptRatio.begin(),ptRatio.end());
+  fillerRatio.fill();
+  iEvent.put(std::move(ptRatioV),"ptRatio");
+
+  std::unique_ptr<edm::ValueMap<float>> ptRelV(new edm::ValueMap<float>());
+  edm::ValueMap<float>::Filler fillerRel(*ptRelV);
+  fillerRel.insert(srcLep,ptRel.begin(),ptRel.end());
+  fillerRel.fill();
+  iEvent.put(std::move(ptRelV),"ptRel");
+
+  std::unique_ptr<edm::ValueMap<float>> jetNDauChargedMVASelV(new edm::ValueMap<float>());
+  edm::ValueMap<float>::Filler fillerNDau(*jetNDauChargedMVASelV);
+  fillerNDau.insert(srcLep,jetNDauChargedMVASel.begin(),jetNDauChargedMVASel.end());
+  fillerNDau.fill();
+  iEvent.put(std::move(jetNDauChargedMVASelV),"jetNDauChargedMVASel");
+
+  std::unique_ptr<edm::ValueMap<reco::CandidatePtr>> jetForLepJetVarV(new edm::ValueMap<reco::CandidatePtr>());
+  edm::ValueMap<reco::CandidatePtr>::Filler fillerjetForLepJetVar(*jetForLepJetVarV);
+  fillerjetForLepJetVar.insert(srcLep,jetForLepJetVar.begin(),jetForLepJetVar.end());
+  fillerjetForLepJetVar.fill();
+  iEvent.put(std::move(jetForLepJetVarV),"jetForLepJetVar");
+
+
+}
+
+template <typename T>
+std::tuple<float,float,float>
+LeptonJetVarProducer<T>::calculatePtRatioRel(edm::Ptr<reco::Candidate> lep, edm::Ptr<pat::Jet> jet, const reco::Vertex &vtx) const {
+ 
+  auto rawp4_ = jet->correctedP4("Uncorrected");
+  auto rawp4 = TLorentzVector(rawp4_.pt(),rawp4_.eta(),rawp4_.phi(),rawp4_.energy());
```

Comment:
Hi @gpetruc - is there a good reason for these type conversions?

Reply:
will be re-implemented avoiding type conversions